driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -326,14 +330,44 @@ private void logServerSelectionFailure(final ServerSelector serverSelector, fina
         }
     }
 
+    @Nullable
+    private ServerTuple selectServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription) {
+        return selectServer(serverSelector, clusterDescription, this::getServer);
+    }
 
-    private ServerTuple selectRandomServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription) {
-        List<ServerDescription> serverDescriptions = serverSelector.select(clusterDescription);
-        if (!serverDescriptions.isEmpty()) {
-            return getRandomServer(new ArrayList<ServerDescription>(serverDescriptions));
-        } else {
-            return null;
+    @Nullable
+    @VisibleForTesting(otherwise = PRIVATE)
+    static ServerTuple selectServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription,
+            final Function<ServerAddress, Server> serverCatalog) {
+        return atMostNRandom(new ArrayList<>(serverSelector.select(clusterDescription)), 2, serverDescription -> {
+            Server server = serverCatalog.apply(serverDescription.getAddress());
+            return server == null ? null : new ServerTuple(server, serverDescription);
+        }).stream()
+                .min(comparingInt(serverTuple -> serverTuple.getServer().operationCount()))
+                .orElse(null);
+    }
+
+    /**
+     * Returns a new {@link List} of at most {@code n} elements, where each element is a result of
+     * {@linkplain Function#apply(Object) applying} the {@code transformer} to a randomly picked element from the specified {@code list},
+     * such that no element is picked more than once. If the {@code transformer} produces {@code null}, then another element is picked
+     * until either {@code n} transformed non-{@code null} elements are collected, or the {@code list} does not have
+     * unpicked elements left.
+     * <p>
+     * Note that this method may reorder the {@code list}.
+     */
+    private static List<ServerTuple> atMostNRandom(final ArrayList<ServerDescription> list, final int n,
```

Comment:
This is very clever and took me some time to fully appreciate the logic.

Reply:
I wish) @jyemin found this algorithm (https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle) mentioned in stackoverflow.com. See the comment https://github.com/mongodb/mongo-java-driver/pull/784#discussion_r713017673.

Reply:
Could you add the algorithm name to the javadoc?

Reply:
Done.